src/app/shared/services/dynamic-data/dynamic-data.service.ts

```diff
@@ -0,0 +1,196 @@
+import { Injectable } from "@angular/core";
+import { addRxPlugin, MangoQuery, RxDocument } from "rxdb";
+import { firstValueFrom, map } from "rxjs";
+
+import { FlowTypes } from "packages/data-models";
+import { environment } from "src/environments/environment";
+import { AppDataService } from "../data/app-data.service";
+import { AsyncServiceBase } from "../asyncService.base";
+import { arrayToHashmap, deepMergeObjects } from "../../utils";
+import { PersistedMemoryAdapter } from "./adapters/persistedMemory";
+import { ReactiveMemoryAdapater, REACTIVE_SCHEMA_BASE } from "./adapters/reactiveMemory";
+import { TemplateActionRegistry } from "../../components/template/services/instance/template-action.registry";
+import { TopLevelProperty } from "rxdb/dist/types/types";
+
+type IDocWithID = { id: string };
+
+@Injectable({ providedIn: "root" })
+/**
+ * The dynamic data service handles the process of loading data from sheets, storing
+ * in memory as a reactive data object, and enabling a persisted write-layer on top
+ * for storing user changes
+ */
+export class DynamicDataService extends AsyncServiceBase {
+  /**
+   * The main DB stores in-memory the rows of any requested sheets, merged with user writes
+   *
+   * Each flow is represented in its own collection, and populated as requested.
+   * This allows users to query and subscribe to data changes in an efficient way
+   */
+  private db: ReactiveMemoryAdapater;
+
+  /**
+   * A separate cache stores user edits flow data, initially in memory
+   * for faster writes but mirrored to a persisted IndexedDB regularly
+   *
+   * This means that user changes will persist app reloads and also be retained
+   * if the underlying data changes so long as compatible (e.g. data list ids still exist)
```

Comment:
This is a good feature, but could it cause issues when app updates introduce data_lists that are compatible with existing user overwrites, but aren't intended to be? I'm not sure I can think of an example, but flagging here more out of interest

Reply:
Agreed, and right now I don't have a specific resolution code or tests to cover those cases. 

Currently, we just have a fairly basic deep merge, where the persisted saved data is merged on top of the sheets. It would make sense to customise the functionality to ensure that the data_list row exists before trying to merge on top and not just leaving dangling rows.

I've added a placeholder test and will update the todo above to cover it.